### PR TITLE
[Small Feature] OpenGR and PointMatcher wrappers

### PR DESCRIPTION
**Pull request by necipfazil**
## Summary

Adds wrappers for OpenGR and pointmatcher libraries.

Both wrappers allow computing registration transformation of two CGAL point sets, where OpenGR wrapper uses Super4PCS algorithm while pointmatcher wrapper uses ICP algorithm.

OpenGR (Super4PCS) does a coarse registration while pointmatcher (ICP) does a fine registration when compared. Therefore, those two wrappers together can be pipelined to have precise registration results in an efficient manner. 

TODO:

- [x] Changes.md to 5.1
- [ ] news entry (?)

## Release Management

* Affected package(s): Point set processing
* Small Feature:  [Features/Small_Features/OpenGR_and_PointMatcher_wrappers](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/OpenGR_and_PointMatcher_wrappers)
* Link to compiled documentation: [Documentation/Point_set_processing](http://www.necipyildiran.com/other/cgal_opengr_pointmatcher_wrappers_doc/Point_set_processing_3/)
* License and copyright ownership: [CLA-GF](https://www.cgal.org/CLA/CLA-GF.html), GeometryFactory
* Developed and locally tested (GNU/Linux) with CGAL 5.0

* GSoC: https://cgal.geometryfactory.com/CGAL/Members/wiki/GSoC2019/OpenGR

### CODE REVIEW

**Comment by afabri** [comment]
I don't get pointmatcher compiled with VC++ and Eigen master.   I filed an [issue](https://github.com/ethz-asl/libpointmatcher/issues/352).  But maybe it is a problem in Eigen @ggael

**Comment by afabri** [comment]
@necipfazil @sgiraudot @nmellado what are the 3rdparty libs in OpenGR about, and do I need them for the examples in this PR?

**Comment by sgiraudot** [comment]
@necipfazil I did not have the time to do a full review yet but I do have two remarks for now:

- The [aggregate initialization with attribute naming](https://en.cppreference.com/w/cpp/language/aggregate_initialization) that you used in several places (`const ICP_config null_config { .name = "_null_pm_config_in_cgal" }` for example) is a C++20 feature. CGAL only support up to C++14, so you should remove them and use other types of initialization.
- Your branch is outdated, you should rebase it on `cgal/master`. In particular, named parameters have changed a bit, `boost::choose_param` is now `parameters::choose_parameter`, `boost::get_param` is now `parameters::get_parameter` and `cgal_bgl_named_params` is not `Named_function_parameters`. I already fixed that locally, if you give me writing access on your CGAL fork, I can push them if you want

**Comment by necipfazil** [comment]
> @necipfazil I did not have the time to do a full review yet but I do have two remarks for now:
> 
>     * The [aggregate initialization with attribute naming](https://en.cppreference.com/w/cpp/language/aggregate_initialization) that you used in several places (`const ICP_config null_config { .name = "_null_pm_config_in_cgal" }` for example) is a C++20 feature. CGAL only supports up to C++14, so you should remove them and use other types of initialization.
Solved: 218286317e56d00aa25dc8b908f8ef6c40df36a5
> 
>     * Your branch is outdated, you should rebase it on `cgal/master`. In particular, named parameters have changed a bit, `boost::choose_param` is now `parameters::choose_parameter`, `boost::get_param` is now `parameters::get_parameter` and `cgal_bgl_named_params` is not `Named_function_parameters`. I already fixed that locally, if you give me writing access on your CGAL fork, I can push them if you want
I sent the invitation for collabration. Thanks!

**Comment by sgiraudot** [comment]
Ok, I did the following things:

1. Rebased the branch on `cgal/master` and fixed conflicts
2. Updated the registration plugin in the demo to include `pointmatcher`. To do something handy, I added an undocumented NP to the `pointmatcher` wrappers so that we can load a config file. This should not break anything.

**Comment by sgiraudot** [comment]
@necipfazil Oops, I think I overwritten your fixes!

Can you please push them back? Be careful when you pull, you might overwrite your changes locally. I think a `git pull --rebase` should do.

**Comment by lrineau** [comment]
[**@sgiraudot**](https://github.com/sgiraudot) commented on [Sep 2, 2019, 10:09 AM GMT+2](https://github.com/CGAL/cgal/pull/4153#issuecomment-527047733 "2019-09-02T08:09:04Z - Replied by Github Reply Comments"):
> @necipfazil Oops, I think I overwritten your fixes!
> 
> Can you please push them back? Be careful when you pull, you might overwrite your changes locally. I think a `git pull --rebase` should do.

As [Github says in the discussion][1], the reference before your force-push was https://github.com/CGAL/cgal/commit/218286317e56d00aa25dc8b908f8ef6c40df36a5. Everything can be recovered.

[1]: https://github.com/CGAL/cgal/pull/4153#event-2600232153

**Comment by sgiraudot** [comment]
@necipfazil nevermind, I fixed my mess thanks to @lrineau :)

File: Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt [hunk 1]
```diff
@@ -314,6 +314,296 @@ points in the domain.
...
+
+The following example reads two point sets and aligns them using the 
+\ref thirdpartylibpointmatcher library, using the ICP algorithm. It also shows how
+to customize ICP algorithm by using possible configurations:
+\cgalExample{Point_set_processing_3/registration_with_pointmatcher.cpp}
+
+\cgalFigureRef{Point_set_processing_3tableRegistrationRegistration_visualization_table} demonstrates
+visualization of a scan data before and after different registration methods are applied,
+including the PointMatcher registration method. To obtain the results for PointMatcher registration
+in the visualization table, abovementioned example was used.
```
Comment (sgiraudot) [inline Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt]: abovementioned -> above-mentioned

File: Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt [hunk 2]
```diff
@@ -314,6 +314,296 @@ points in the domain.
...
+
+The chain of filters to be applied to the point cloud. The
+point cloud is processed into an intermediate point cloud with the given chain
+of filters to be used in the alignment procedure. The chain is organized with
+the forward traversal order of the point set filters range.
+
+The chain of point set filters are applied only once at the beginning of the
+ICP procedure, i.e., before the first iteration of the ICP algorithm.
+
+The filters can have several purposes, including but are not limited to:
```
Comment (sgiraudot) [inline Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt]: including but are not limited to -> including but not limited to

File: Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt [hunk 3]
```diff
@@ -314,6 +314,296 @@ points in the domain.
...
+
+- removal of noisy points which render alignment of point clouds difficult,
+
+- removal of redundant points so as to speed up alignment,
+
+- addition of descriptive information to the points such as a surface normal vector or the direction from the point to the sensor.
+
+In registration, there are two point clouds in consideration, one of which is the reference point
+cloud while the other one is the point cloud that a transformation is computed to register that point
+cloud to the reference. The point set filters corresponds to `readingDataPointsFilters` configuration module of \ref thirdpartylibpointmatcher
```
Comment (sgiraudot) [inline Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt]: while the other one is the point cloud that a transformation is computed to register that point
cloud to the reference -> while the other one is the point cloud to register.

File: Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt [hunk 4]
```diff
@@ -314,6 +314,296 @@ points in the domain.
...
+It depicts a use case where a coarse estimation of a registration transformation
+is done using the Super4PCS algorithm. Then, a fine registration from this coarse
+registration using the ICP algorithm. 
+\cgalExample{Point_set_processing_3/registration_with_opengr_pointmatcher_pipeline.cpp}
+
+\cgalFigureRef{Point_set_processing_3tableRegistrationRegistration_visualization_table} demonstrates
+visualization of a scan data before and after different registration methods are applied,
+including the pipeline of %OpenGR and PointMatcher registration methods. To obtain the results
+for the pipeline of %OpenGR and PointMatcher registration methods in the visualization table,
+abovementioned example was used.
```
Comment (sgiraudot) [inline Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt]: abovementioned -> above-mentioned

File: Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt [hunk 5]
```diff
@@ -314,6 +314,296 @@ points in the domain.
...
+\subsection Point_set_processing_3Examples_registration_OpenGR OpenGR Example
+
+The following example reads two point sets and aligns them using the 
+\ref thirdpartyOpenGR library, using the Super4PCS algorithm:
+\cgalExample{Point_set_processing_3/registration_with_OpenGR.cpp}
+
+\cgalFigureRef{Point_set_processing_3tableRegistrationRegistration_visualization_table} demonstrates
+visualization of a scan data before and after different registration methods are applied,
+including the %OpenGR registration method. To obtain the results for %OpenGR registration
+in the visualization table, abovementioned example was used.
```
Comment (sgiraudot) [inline Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt]: abovementioned -> above-mentioned

**Comment by necipfazil** [comment]
> @necipfazil I made a review, there are several minor things to correct in the text, you can fix them whenever you have the change.
> 
> Apart from that, it looks good to me. I pre-approve this feature.
> 
> Nice work!

The flaws pointed in the review were resolved: https://github.com/CGAL/cgal/pull/4153/commits/d53a54ca1cae5e7b03ec3699b8b587f07bff4a1a

Thanks!

**Comment by sgiraudot** [comment]
Thanks @necipfazil for the corrections!

For information, I just added a couple of things to this feature:
- a named parameter `maximum_angle_deviation` to set the corresponding parameter in _OpenGR_
- warnings about which point set should be used as reference with _OpenGR_ in the case where a small and big point sets are used (as it seems counter-intuitive)

**Comment by sgiraudot** [comment]
- This small feature is approved
- `CHANGES.md` is up to date
- Conflicts have been fixed
- The PR is now ready to be tested

@maxGimeno could you please install [libpointmatcher](https://github.com/ethz-asl/libpointmatcher/) and [OpenGR](https://github.com/STORM-IRIT/OpenGR/) on Travis an some test platforms for this PR to be fully tested? Thanks!

**Comment by sloriot** [comment]
Branch weights 2.4Mo. I think `hippo1.xyz` and `hippo2.xyz` are responsible for that (1.3Mo and 887Ko respectively once gzipped). Shall we only ship a smaller version of those files and refer to the larger one from OpenGR repo? I mean OpenGR already has it own testsuite so there is no need that we test the outcome of the algo directly in CGAL.

**Comment by maxGimeno** [comment]
@sgiraudot It would be too heavy for travis. I will not impact every run of the CI to test 3 examples. But I'll install them on some machines of the testsuite.

**Comment by maxGimeno** [comment]
Error in travis : https://travis-ci.org/CGAL/cgal/jobs/637328224#L5724

**Comment by sgiraudot** [comment]
The dependencies were not handled properly, I introduce a mechanism similar to what I did in https://github.com/CGAL/cgal/pull/4356 for OpenGR and pointmatcher.

**Comment by sloriot** [comment]
I just pushed-force the filtered branch having a size of 664K. I also create an [archive tag](https://github.com/CGAL/cgal-public-dev/releases/tag/archive%2Fnecipfazil%2Fgsoc2019-pointmatcher_icp_wrapper-necipfazil_before_filtering). You have to have reset your local branches after a fetch.

File: Point_set_processing_3/examples/Point_set_processing_3/CMakeLists.txt [hunk 6]
```diff
@@ -78,6 +78,34 @@ if ( CGAL_FOUND )
...
+
+    # Executables that require both libpointmatcher and OpenGR
+    if (libpointmatcher_FOUND AND OpenGR_FOUND)
+      create_single_source_cgal_program( "registration_with_opengr_pointmatcher_pipeline.cpp" )
+      CGAL_target_use_OpenGR(registration_with_opengr_pointmatcher_pipeline)
+      CGAL_target_use_pointmatcher(registration_with_opengr_pointmatcher_pipeline)
+    else()
+      message(STATUS "NOTICE: registration_with_opengr_pointmatcher_pipeline requires libpointmatcher and OpenGR, and will not be compiled.")
+    endif()
+
```
Comment (maxGimeno) [inline Point_set_processing_3/examples/Point_set_processing_3/CMakeLists.txt]: You need to link with Eigen for pointmatcher and OpenGR stuffs.

**Comment by maxGimeno** [comment]
Also there are Cmake warnings if pointmatcher is not found, it should not: https://cgal.geometryfactory.com/CGAL/testsuite/#Point_set_processing_3_Examples

File: Point_set_processing_3/examples/Point_set_processing_3/CMakeLists.txt [hunk 7]
```diff
@@ -78,6 +78,34 @@ if ( CGAL_FOUND )
     create_single_source_cgal_program( "jet_smoothing_example.cpp" )
     create_single_source_cgal_program( "normal_estimation.cpp" )
     create_single_source_cgal_program( "edges_example.cpp" )
+    
+    # Executables that require libpointmatcher
+    find_package(libpointmatcher)
```
Comment (maxGimeno) [inline Point_set_processing_3/examples/Point_set_processing_3/CMakeLists.txt]: add QUIET to avoid unwanted warnings

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-88.shtml

**Comment by lrineau** [comment]
@sloriot @sgiraudot The branch weights 660kB. Is that right?

**Comment by sloriot** [comment]
It was 664K after my rebase so it looks good.

**Comment by lrineau** [comment]
This PR conflicts with `master`.

**Comment by sgiraudot** [comment]
Rebased.

**Comment by sloriot** [comment]
[Error](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-98/Point_set_processing_3_Examples/TestReport_Friedrich_Ubuntu-gcc-7.gz)

**Comment by maxGimeno** [comment]
[warnings](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-100/Point_set_processing_3_Examples/TestReport_Friedrich_Ubuntu-gcc-7.gz)

**Comment by sgiraudot** [comment]
Warnings on our side should be fixed. I did a [fix for the OpenGR side](https://github.com/STORM-IRIT/OpenGR/pull/46) but we'll have to wait for its integration.

**Comment by maxGimeno** [comment]
Thanks, but there are now conflicts with master

**Comment by sloriot** [comment]
> Thanks, but there are now conflicts with master

Not a big deal, I have the rerere locally.

**Comment by maxGimeno** [comment]
conflicts with #4418 that removes GetPointMap from PSP

**Comment by maxGimeno** [comment]
Successfully tested by https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-104.shtml

**Comment by sloriot** [comment]
Thanks a lot guys for the long term effort. It is now part of the master branch. Congrats @necipfazil !

cc @nmellado.

**Comment by sloriot** [comment]
@sgiraudot  is taking care of the web entry.

**Comment by lrineau** [comment]
@sloriot You merged the PR, but commit https://github.com/CGAL/cgal/pull/4153/commits/70f3ac32efd29fadc871be99adcdee2d882c94d4 was not tested.

**Comment by sloriot** [comment]
I know I asked the doc to be fixed.

**Comment by nmellado** [comment]
Question regarding further updates on OpenGR side, do you see any problem having OpenGR requesting c++17 features ?

File: Installation/cmake/modules/CGAL_target_use_OpenGR.cmake [hunk 8]
```diff
@@ -0,0 +1,9 @@
+if (CGAL_target_use_OpenGR_included)
+  return()
+endif()
+set(CGAL_target_use_OpenGR_included TRUE)
+
+function(CGAL_target_use_OpenGR target)
+  target_include_directories(${target} PUBLIC ${OpenGR_INCLUDE_DIR})
+  target_compile_options( ${target} PUBLIC -DCGAL_LINKED_WITH_OPENGR)
+endfunction()
```
Comment (lrineau) [inline Installation/cmake/modules/CGAL_target_use_OpenGR.cmake]: @sgiraudot Why not link with the imported targets defined by `OpenGRConfig.cmake`?

That would solve the requirement for C++17: CGAL targets using OpenGR would be compiled with C++17 enabled (and automatically, using CMake features).

Cc: @nmellado

Comment (nmellado) [reply Installation/cmake/modules/CGAL_target_use_OpenGR.cmake]: This should be indeed exported, I made changes for that on `master` (https://github.com/STORM-IRIT/OpenGR/commit/ad80cacbdb5cc9bf034e9a9613f1982105e6708d), see https://github.com/STORM-IRIT/OpenGR/blob/7da3b75a6b153ea243f9190fd62a8d189d9ba70f/src/gr/accelerators/CMakeLists.txt#L37-L39.

Note also that the library is exported in a cmake namespace `gr`, so `target_link_library( ... gr::XXX)` should be enough to get your target properly configured (maybe I am missing something here).

**Pull request merged**